### ISSUE TITLE
Secondary Indexes with multiple values?

### ISSUE THREAD

**Issue by huguesbr** [yapstudios/YapDatabase#431]
Hi,

Let's imagine I have an object which look like this: 
```
class Obj {
 let identifier: String
 let values: [String]
 ...
}
```

I want to be able to find all my objects containing a specific values.

If I use Secondary Indexes, I can't return multiple index for the same object and can't set array for a specific index.

```
struct ObjIndex {
  static let indexName = "objIndex"
  static let key = "values"
  
  static func register(database: YapDatabase) {
    let setup = YapDatabaseSecondaryIndexSetup()
    setup.addColumn(key, with: .text)
    let handler = YapDatabaseSecondaryIndexHandler.withObjectBlock { (transaction, dict, collection, key, obj) in
      guard let conversation = obj as? Obj else { return }
      dict[key] = obj.values // invalid will result in NULL in the index column
      // also impossible to return multiple dict with each values
    }
    let index = YapDatabaseSecondaryIndex.init(setup: setup, handler: handler)
    database.register(index, withName: index)
  }
}
```

Any suggestions, ideas?

**Comment by chrisballinger** [comment]
You you'll need to serialize the .values array to Data and deserialize on the other side. You can use a data blob column type.

**Comment by huguesbr** [comment]
@chrisballinger thx but can I then search using the blob?

`YapDatabaseQuery(string: "WHERE values contains ?", parameters: [myValue])`

I don't see how I will search in my opaque blob data directly in SQLite. If I need to deserialize, then I will loose the benefits of the index?

**Comment by xquezme** [comment]
Are you trying [FTS](https://github.com/yapstudios/YapDatabase/wiki/Full-Text-Search)?

**Comment by huguesbr** [comment]
@xquezme not sure it’s adapted as it’s not text but ids in my case ( I’ve tweak the sample code for the issue ). But I could if no alternative. I was hoping for a way to index multiple items for the list of matching values or index an array of values for the same items. 
I will try to work around with `fts` if no better ways, thank you.

**Comment by chrisballinger** [comment]
@huguesbr I believe you can do partial matches on blobs, with some limitations, but a hack with FTS might be easier.

Can you explain your exact use case? Is this for looking up database relationships between objects?

**Comment by huguesbr** [comment]
yes it’s a relationship but that is not express in the dB. 
I have a message which have multiple author and want to fetch all the message concerning  specific author. Authors are not store in a different table/view they are store on the message’s object itself. 
Maybe I should explore relationships ( i have to say the notion of edges kind of restrain me from going further... :) or maybe storing metadata on the message.

**Comment by huguesbr** [comment]
I want to add something unrelated but I have been working for only couple of weeks with YapDatabase and you’ve done a terrific work to adapt database concepts to local/mobile usage, specially with the notion of dispatch data in bucket using view ( “slow” to write / blazing fast to read ). Cheers to that

**Comment by chrisballinger** [comment]
Yeah I think this would be better handled with a different model design. It's possible to have many-to-many relationships between authors and messages. See https://github.com/yapstudios/YapDatabase/wiki/Relationships

It might seem complicated at first but it's really not so bad (just implement [YapDatabaseRelationshipNode protocol](https://github.com/yapstudios/YapDatabase/wiki/Relationships#yapdatabaserelationshipnode-protocol) on your objects). I'm pretty sure this is what you're looking for. The only caveat is that if you want to do further lookups/searches beyond just relationships, you'll still need some sort of secondary index/fts stuff.

**Comment by huguesbr** [comment]
Ok, you’ve convinced me to not fight the system. I will take a break and come back to it using relationships, as I will probably need it for some `filteredViews` later down the road. Thank you.
